Content.Server/Internals/InternalsProvider.cs

```diff
@@ -0,0 +1,161 @@
+
+        public void OnContainerRemoved(EntityUid uid, InternalsProviderComponent component, EntRemovedFromContainerMessage args)
+            => DisconnectFromInternals(uid, component);
+
+        public void OnToggleInternals(EntityUid uid, InternalsProviderComponent component, ToggleInternalsEvent args)
+        {
+            if (args.BypassCheck == false)
+            {
+                var user = GetInternalsComponent(component)?.Owner; // Todo: grab from event
+                if (user == null || !Get<ActionBlockerSystem>().CanUse(user))
+                    return;
+            }
+
+            //Logger.DebugS("internals", $"Trying to set internals @{uid} to \"{args.ForcedState ?? !component.Connected}\" - {(args.ForcedState != null ? "Forced ": "")}{(args.BypassCheck ? "Bypassed!" : "")}");
+
+            var wasConnected = component.Connected;
+            if (args.ForcedState == true || (args.ForcedState == null && !component.Connected))
+                ConnectToInternals(uid, component);
+            else if (args.ForcedState == false || (args.ForcedState == null && component.Connected))
+                DisconnectFromInternals(uid, component);
+
+            args.Handled = wasConnected != component.Connected || component.Connected == args.ForcedState;
+
+            if (args.Actions is not null)
+                args.Actions.GrantOrUpdate(ItemActionType.ToggleInternals, GetInternalsComponent(component) != null, component.Connected);
+            else
+            {
+                ItemActionsComponent? actions = null;
+                if (Resolve(uid, ref actions))
+                    actions.GrantOrUpdate(ItemActionType.ToggleInternals, GetInternalsComponent(component) != null, component.Connected);
+            }
+
+            if (args.BypassCheck && !args.Handled)
+                Logger.ErrorS("internals", $"Internals @{uid} state change failed to change to \"{args.ForcedState}\" despite being forced to bypass a usability check");
+        }
+
+        public void ConnectToInternals(EntityUid uid, InternalsProviderComponent component)
+        {
+            if (component.Connected) return;
+
+            var internals = GetInternalsComponent(component);
+            if (internals?.TryConnectTank(component.Owner) ?? false)
+                component.Internals = internals;
+        }
+
+        public void DisconnectFromInternals(EntityUid uid, InternalsProviderComponent component)
+        {
+            if (!component.Connected) return;
+
+            //Logger.DebugS("internals", $"Disconnecting internals @{uid} from {component.Internals!.Owner.Name}");
+
+            var internals = component.Internals!;
+            component.Internals = null;
+            internals.DisconnectTank(true);
+
+            ItemActionsComponent? actions = null;
+            if (Resolve(uid, ref actions))
+                actions.GrantOrUpdate(ItemActionType.ToggleInternals, GetInternalsComponent(component) != null, component.Connected);
```

Comment:
You should change this so actions can be passed in; this will also log errors if actions don't exist which I'm not sure is intended.